spring-kafka/src/main/java/org/springframework/kafka/annotation/KafkaListenerAnnotationBeanPostProcessor.java

```diff
@@ -690,6 +696,26 @@ protected boolean isEmptyPayload(Object payload) {
 			return defaultFactory;
 		}
 
+		private void addFormatters(FormatterRegistry registry) {
+			for (Converter<?, ?> converter : getBeansOfType(Converter.class)) {
+				registry.addConverter(converter);
+			}
+			for (GenericConverter converter : getBeansOfType(GenericConverter.class)) {
+				registry.addConverter(converter);
+			}
+			for (Formatter<?> formatter : getBeansOfType(Formatter.class)) {
+				registry.addFormatter(formatter);
+			}
+		}
+
+		private <T> Collection<T> getBeansOfType(Class<T> type) {
+			if (KafkaListenerAnnotationBeanPostProcessor.this.beanFactory instanceof ListableBeanFactory) {
+				return ((ListableBeanFactory) KafkaListenerAnnotationBeanPostProcessor.this.beanFactory).getBeansOfType(type).values();
```

Comment:
I checked the call stack and all of these `getBeansOfType()` is not good during `BeanPostProcessor.postProcessAfterInitialization()`. That is an "early BeanFactory initialization syndrome".

If we are going this `add*` way, we should consider to deffer their population until `afterSingletonsInstantiated()` phase.